articles/qos_configurability.md

```diff
@@ -0,0 +1,379 @@
+typename PublisherT = rclcpp::Publisher<MessageT, AllocatorT>>
+std::shared_ptr<PublisherT>
+create_publisher(
+  const std::string & topic_name,
+  const rclcpp::QoS & qos,
+  const rclcpp::QosOverridingOptions & qos_options = QosOverridingOptions{false},
+  const PublisherOptionsWithAllocator<AllocatorT> & options = PublisherOptionsWithAllocator<AllocatorT>()
+);
+```
+
+```cpp
+// example usage
+node->create_publisher(
+    "my_topic",
+    user_qos_provided_in_code_1,
+    rclcpp::KeepLast(10),
+    {  // implicit, to lower verbosity
+        {QosPolicyKind::Reliability, QosPolicyKind::History, QosPolicyKind::HistoryDepth}, // only declare parameters for history depth, history kind, reliability
+        [](const QoSProfile qos) -> bool {
+            return qos.is_keep_all() || qos.history_depth() > 10u;  // constrains involving more than one QoS in callbacks
+        },
+        "my_id",  // id to disambiguate entities in same topic
+    });
+
+// nothing is reconfigurable
+node->create_publisher(
+    "my_topic",
+    user_qos_provided_in_code_2,
+    rclcpp::KeepLast(10));
+
+// allow reconfiguring durability
+node->create_publisher(
+    "another_topic",
+    user_qos_provided_in_code_3,
+    rclcpp::KeepLast(10),
+    QosOverridingOptions{QosPolicyKind::Durability});
+```
+
+The above API mentions having some "defaults". 
+The intend of that is to provide a less verbose option, in which case the user can opt-in qos reconfigurability by only opting-in a flag.
+
+Proposed defaults:
+- Reliability
+- History kind
+- History depth
+- Durability
+
+TBD: Do we want defaults?
+
+## Reconfigurability of topics the ROS 2 core creates
+
+TBD: discuss reconfigurability of topics packages in the ROS 2 core creates here, e.g.: clock, rosout, tf, tf_static, etc.
+
+## Further enhancements
+
+### Reusing profiles
+
+The above proposal doesn't include a mechanism to reuse the same qos profile in the parameter files.
+We could, for example, leverage yaml anchors to allow this:
+
+```yaml
+/my/ns/node_name:
+  ros__parameters:
+    qos_profiles:
+      publisher: &profileA
+        'my/fully/qualified/topic/name/here': 
+          reliability: reliable
+          history_depth: 1
+          history: keep_last
+/my/ns/node_name:
+  ros__parameters:
+    qos_profiles:
+      subscription:
+        'my/fully/qualified/topic/name/here': 
+          <<: *profileA,
+          history_depth: 100  # override the history depth
+```
+
+### Extending the ROS qos profile with a rmw vendor specific payload
+
+TBD
+
+## QoS resolution order
+
+- The QoS provided by the user in the publisher/subscription constructor is used as a default.
+- Those defaults are overridden with the values provided in parameters, if the user allowed this kind of reconfigurability.
+- If the policy was set to `RMW_QOS_*_POLICY_DEFAULT`, the qos resolution mechanism provided by the rmw vendor will be applied (e.g. DDS XML profile files). If not, the previous policy value is the final one.
```

Comment:
We should probably clarify this point. Are you referring to the policy set *after* resolving any parameter overrides or *before*? IMO, this should happen after.

Reply:
Yes, it should happen after. I will rephrase.

Reply:
Rephrased https://github.com/ros2/design/pull/300/commits/9f5f17f94b009251b2da560daedf0f383f8d32fc.